File: server/app/schemas/analytics.py
"""Analytics response models (spec section 5.6).

OEE is always reported with its three terms alongside it. Spec section 5.6 is
explicit about why: "Display the three components separately so users can
understand why OEE changes." A single OEE number tells a manager something moved
but not whether to look at breakdowns, cycle times or scrap.
"""

from __future__ import annotations

from datetime import date
from uuid import UUID

from pydantic import BaseModel, Field


class OEEComponents(BaseModel):
    """The three OEE terms and their product.

    Each term is capped at 100%. Performance is the one that can genuinely
    exceed it, and a value above 100% means the recorded ideal cycle time is
    wrong rather than that a machine beat physics -- `performance_uncapped`
    keeps that visible instead of hiding it behind the cap.
    """

    availability_percentage: float = Field(
        ge=0, le=100, description="Operating time / planned production time."
    )
    performance_percentage: float = Field(
        ge=0, le=100, description="Actual output / ideal output for the operating time."
    )
    quality_percentage: float = Field(ge=0, le=100, description="Good units / total units.")
    oee_percentage: float = Field(ge=0, le=100, description="Availability x Performance x Quality.")
    performance_uncapped_percentage: float = Field(
        ge=0,
        description=(
            "Performance before capping. Above 100 indicates an ideal cycle time "
            "shorter than the machine's real capability -- a reference-data problem."
        ),
    )


class OEESummary(OEEComponents):
    """OEE across a filtered period, with the inputs it came from."""

    start_date: date
    end_date: date

    operating_minutes: int = Field(ge=0)
    planned_minutes: int = Field(ge=0)
    downtime_minutes: int = Field(ge=0)
    produced_quantity: int = Field(ge=0)
    accepted_quantity: int = Field(ge=0)
    rejected_quantity: int = Field(ge=0)
    ideal_output: float = Field(
        ge=0, description="Units producible in the operating time at rated cycle speed."
    )

    record_count: int = Field(ge=0)
    has_data: bool = Field(description="False when no production matched the filter.")


class OEETrendPoint(OEEComponents):
    """One day on the OEE trend chart."""

    bucket_date: date


class OEEByMachine(OEEComponents):
    """OEE for one machine, for fleet comparison."""

    machine_id: UUID
    machine_code: str
    machine_name: str
    produced_quantity: int = Field(ge=0)
    operating_minutes: int = Field(ge=0)


class EfficiencySummary(BaseModel):
    """Production efficiency against plan and target."""

    start_date: date
    end_date: date

    total_planned: int = Field(ge=0, description="Sum of planned quantities on production runs.")
    total_produced: int = Field(ge=0)
    total_target: int = Field(ge=0, description="Sum of daily targets from daily_targets.")

    efficiency_percentage: float = Field(ge=0, description="Produced / planned.")
    achievement_percentage: float = Field(ge=0, description="Produced / target.")

    total_planned_minutes: int = Field(ge=0)
    total_operating_minutes: int = Field(ge=0)
    total_downtime_minutes: int = Field(ge=0)
    downtime_percentage: float = Field(
        ge=0, le=100, description="Downtime as a share of planned production time."
    )

    record_count: int = Field(ge=0)
    has_data: bool


class EfficiencyTrendPoint(BaseModel):
    """One day on the efficiency trend chart."""

    bucket_date: date
    planned_quantity: int = Field(ge=0)
    produced_quantity: int = Field(ge=0)
    target_quantity: int = Field(ge=0)
    efficiency_percentage: float = Field(ge=0)
    achievement_percentage: float = Field(ge=0)


class DowntimeByMachine(BaseModel):
    """Downtime attributed to one machine."""

    machine_id: UUID
    machine_code: str
    machine_name: str
    downtime_minutes: int = Field(ge=0)
    planned_minutes: int = Field(ge=0)
    downtime_percentage: float = Field(ge=0, le=100)


class DefectAnalytics(BaseModel):
    """Defect analysis over a period: Pareto, trend and worst offenders."""

    start_date: date
    end_date: date

    total_inspected: int = Field(ge=0)
    total_rejected: int = Field(ge=0)
    defect_rate_percentage: float = Field(ge=0, le=100)

    # Imported lazily in the module that builds this to avoid a cycle; declared
    # here with forward references resolved at import time in `__init__`.
    by_defect: list[DefectSummary] = Field(
        default_factory=list, description="Pareto-ordered defect breakdown."
    )
    by_machine: list[QualityByDimension] = Field(
        default_factory=list, description="Rejection rate by machine, worst first."
    )
    by_component: list[QualityByDimension] = Field(
        default_factory=list, description="Rejection rate by component, worst first."
    )
    trend: list[DefectTrendPoint] = Field(
        default_factory=list, description="Daily defect rate over the period."
    )
    has_data: bool


from app.schemas.quality import (  # noqa: E402 - resolves the forward references above
    DefectSummary,
    DefectTrendPoint,
    QualityByDimension,
)

DefectAnalytics.model_rebuild()

File: client/types/analytics.ts
/**
 * Analytics response models, mirroring `app/schemas/analytics.py`.
 *
 * OEE always arrives with its three terms. Spec section 12 forbids duplicating
 * the calculation here, and spec section 5.6 requires the terms be shown
 * separately -- a single OEE number says something moved, not whether to look
 * at breakdowns, cycle times or scrap.
 */
import type { DateRangeFilters, IsoDate, Uuid } from "@/types/common";
import type { DefectSummary, DefectTrendPoint, QualityByDimension } from "@/types/quality";

/** The three OEE terms and their product. Each capped at 100%. */
export interface OEEComponents {
  availability_percentage: number;
  performance_percentage: number;
  quality_percentage: number;
  oee_percentage: number;
  /**
   * Performance before capping.
   *
   * Above 100 means a recorded ideal cycle time is shorter than the machine can
   * really achieve -- a reference-data problem, not a machine outperforming
   * physics. Surfaced so the problem stays visible rather than hidden by the cap.
   */
  performance_uncapped_percentage: number;
}

/** OEE across a filtered period, with the inputs it came from. */
export interface OEESummary extends OEEComponents {
  start_date: IsoDate;
  end_date: IsoDate;

  operating_minutes: number;
  planned_minutes: number;
  downtime_minutes: number;
  produced_quantity: number;
  accepted_quantity: number;
  rejected_quantity: number;
  ideal_output: number;

  record_count: number;
  has_data: boolean;
}

/** One day on the OEE trend chart. */
export interface OEETrendPoint extends OEEComponents {
  bucket_date: IsoDate;
}

/** OEE for one machine, for fleet comparison. Returned worst first. */
export interface OEEByMachine extends OEEComponents {
  machine_id: Uuid;
  machine_code: string;
  machine_name: string;
  produced_quantity: number;
  operating_minutes: number;
}

/** Production efficiency against plan and target. */
export interface EfficiencySummary {
  start_date: IsoDate;
  end_date: IsoDate;

  total_planned: number;
  total_produced: number;
  total_target: number;

  efficiency_percentage: number;
  /** Zero when the filter includes a machine; targets have no such dimension. */
  achievement_percentage: number;

  total_planned_minutes: number;
  total_operating_minutes: number;
  total_downtime_minutes: number;
  downtime_percentage: number;

  record_count: number;
  has_data: boolean;
}

/** One day on the efficiency trend chart. */
export interface EfficiencyTrendPoint {
  bucket_date: IsoDate;
  planned_quantity: number;
  produced_quantity: number;
  target_quantity: number;
  efficiency_percentage: number;
  achievement_percentage: number;
}

/** Downtime attributed to one machine. Returned worst first. */
export interface DowntimeByMachine {
  machine_id: Uuid;
  machine_code: string;
  machine_name: string;
  downtime_minutes: number;
  planned_minutes: number;
  downtime_percentage: number;
}

/** Full defect analysis: Pareto, by machine, by component, and trend. */
export interface DefectAnalytics {
  start_date: IsoDate;
  end_date: IsoDate;

  total_inspected: number;
  total_rejected: number;
  defect_rate_percentage: number;

  by_defect: DefectSummary[];
  by_machine: QualityByDimension[];
  by_component: QualityByDimension[];
  trend: DefectTrendPoint[];
  has_data: boolean;
}

/** Filters accepted by the analytics endpoints. */
export interface AnalyticsFilters extends DateRangeFilters {
  machine_id?: Uuid;
  component_id?: Uuid;
  line_id?: Uuid;
}
